homeassistant/components/discord/calendar.py

```diff
@@ -0,0 +1,100 @@
+"""Module for using the calendar entity platform."""
+
+from __future__ import annotations
+
+import datetime
+import logging
+from typing import Any
+
+import nextcord
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+
+PARALLEL_UPDATES = 1
+SCAN_INTERVAL = datetime.timedelta(minutes=1)
```

Comment:
This is a pretty tight scan interval for a calendar. Do you have thoughts about that?

I was reading the nextcord library and I can see that it also appears to cache the events fetched inside the Guild object. Does this use that or is it always polling?

Reply:
Using the fetch method always calls the API. Creating a longer interval would homeassistant less responsive to changes but would also save bandwidth. Would a five minute interval make more sense?

Reply:
There are two things to consider here, given my experience with cloud calendars e.g. Google. I also realize there is some interplay with triggers to consider.

What google calendar does is:
(1) polls the API every 15 minutes ([update coordinator](https://github.com/home-assistant/core/blob/20d70337d51c0608d46cdfe17e5d8bd79be33947/homeassistant/components/google/calendar.py#L243)). This means that you don't notice new events for 15 minutes. So if you rely on an automation running as soon as an event is scheduled it won't catch it. I'm not Sure how true that is for this use case, but i assume a guild event would be scheduled far in advance of the event.
(2) The other interesting part is around the `on`/`off` state of the calendar entity. Google calendar run async_update every minute so it notices when the next event fetched in #1 has started. (See [should_poll](https://github.com/home-assistant/core/blob/20d70337d51c0608d46cdfe17e5d8bd79be33947/homeassistant/components/google/calendar.py#L307) which will use the default 1 minute update and see the calendar entity [state update method](https://github.com/home-assistant/core/blob/20d70337d51c0608d46cdfe17e5d8bd79be33947/homeassistant/components/calendar/__init__.py#L283)

So this is a little complicated as it both uses two types of polling.

The last thing to consider is that triggers also have a 15 minute poll interval using `async_get_events`. Triggers get the list of upcoming events in the next 15 minute window, then schedule alarms at the right moment based on when they start, and iterates through all the events.    This may stack up your delays here...  The assumption was that `async_get_events` may actually send an RPC itself, like it does with google calendar, todoist, etc.   (I was thinking of moving google calendar to do something like what you have here to read from a local cache, but haven't done it yet since we need to support search).

Reply:
1) I think you're right about the use case, events typically are scheduled well in advance, so just from that, a longer scan interval would make more sense.

2) Assuming we figure out how to access the status of events, we could do two types of polling here as well. A) We fetch Discord events with `async_update` every 10 minutes and add them locally without removing them. Then have `async_get_events` convert them into homeassistant calendar events. B) I'm `async_get_évents` we poll the status of the events and if the event is active we say the start time of the event is now when converting to homeassistant calendar event objects. If the status is completed we remove the event locally. If the status is scheduled then provide its scheduled start time. One thing I'm wondering is that will saying the start time is now trigger triggers successfully? On that subject, how do calendar triggers work exactly?

Reply:
I don't think you need status anywhere since you said the api filters out completed events. The event is determined to be active based on the start/end time of the returned `CalendarEvent` only.  See [architecture discussion](https://github.com/home-assistant/architecture/discussions/700#discussioncomment-1902322) end of that thread to see how triggers were proposed to work. See `trigger.py` in calendar for the low level details.

Reply:
What I'm saying is that if a user starts an event before the scheduled time in discord, we should change the homeassistant calendar event start time to now instead of it's discord scheduled start time which should trigger calendar triggers.